### PR TITLE
Validation model specs

### PR DESCRIPTION
**Pull request by the-faizmohammad**
This pull request consist of following changes
**Implemented Changes:**

- Added the attribute validations for the User and Post modules.
- User validations: for name attribute used the presence helper, for posts_counter used numericality.
- Post validations: for title used presence, length helpers, for comments_counter and likes_counter used numericality.
- Installed the RSpec to implement unit tests for the existing model classes.
- To match the validations for the counters, updated the default value of counter from nil to 0.
- Implemented test suites for the classes User, Post, Comment and Like classes checking the validations and existing methods such as update_counter methods and recent_posts, recent_comments methods.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+RSpec.describe User, type: :model do
+  it 'is invalid without a name' do
+    user = User.new(name: nil)
+    expect(user).to_not be_valid
+  end
+
+  it 'is valid with a non-negative integer posts counter' do
+    user = User.new(name: 'Alice', posts_counter: 0)
+    expect(user).to be_valid
+  end
```
Comment (omar-muhamad) [inline spec/models/user_spec.rb]: - [ ] Please add another test case that returns `not valid` if the `posts_counter`  has a negative value  to cover all edge cases.
```suggestion
  it 'is valid with a non-negative integer posts counter' do
    user = User.new(name: 'Alice', posts_counter: 0)
    expect(user).to be_valid
  end
  
  it 'is not valid if posts counter has a negative value' do
    user = User.new(name: 'Alice', posts_counter: -2)
    expect(user).to_not be_valid
  end
```

Or It will be much better professionally if you test the validations like this:
```ruby
describe 'validations' do
  it { should validate_presence_of(:name) }
  it { should validate_numericality_of(:posts_counter).greater_than_or_equal_to(0) }
end
```

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+RSpec.describe Post, type: :model do
+  it 'is invalid without a title' do
+    post = Post.new(title: nil, text: 'This is a test post.')
+    expect(post).to_not be_valid
+  end
+
+  it 'is invalid with a title exceeding 250 characters' do
+    post = Post.new(title: 'a' * 251, text: 'This is a test post.')
+    expect(post).to_not be_valid
+  end
```
Comment (omar-muhamad) [inline spec/models/post_spec.rb]: - [ ] You only added validations for post title presence and length, Please also add test cases for the validation of  `likes_counter` and `comments_counter`.

- [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=equal%20to%20zero.-,Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.,-Need%20a%20big), Please add unit tests for `five_recent_comments` and `update_posts_counter` methods because you forgot to add them.

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+RSpec.describe User, type: :model do
+  it 'is invalid without a name' do
+    user = User.new(name: nil)
+    expect(user).to_not be_valid
+  end
+
+  it 'is valid with a non-negative integer posts counter' do
+    user = User.new(name: 'Alice', posts_counter: 0)
+    expect(user).to be_valid
```
Comment (omar-muhamad) [inline spec/models/user_spec.rb]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=equal%20to%20zero.-,Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.,-Need%20a%20big), Please add unit tests for the `three_recent_posts` method because you forgot to add them.

File: spec/models/like_spec.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    post = Post.create(title: 'Test Post', text: 'This is a test post.')
+    like = Like.new(post:)
+    expect(like).to_not be_valid
+  end
+
+  it 'is invalid without a post' do
+    user = User.create(name: 'Liker')
+    like = Like.new(user:)
+    expect(like).to_not be_valid
+  end
```
Comment (omar-muhamad) [inline spec/models/like_spec.rb]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=equal%20to%20zero.-,Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.,-Need%20a%20big), Please add unit tests for the `update_likes_counter` method.

File: spec/models/comment_spec.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    end
+
+    it 'check to create comment without a post to the user' do
+      expect(Comment.new(user: @user)).to_not be_valid
+    end
+
+    it 'check to create comment to a post without a user' do
+      expect(Comment.new(post: @post)).to_not be_valid
+    end
+  end
```
Comment (omar-muhamad) [inline spec/models/comment_spec.rb:21]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=equal%20to%20zero.-,Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.,-Need%20a%20big), Please add unit tests for the `update_comments_counter` method.

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @the-faizmohammad,

### Nice job, you are almost there.

cat-driving-serious

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by shahadat3669 — APPROVED** [review]
Hi @the-faizmohammad, 

Your project is complete!  There is nothing else to say other than it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

✅ No linter Errors
✅ Testing was exceptional 👍 
✅ The code is top-notch! 💯
**Your testing was exceptional. The comprehensive coverage of diverse input scenarios truly impressed me. Thank you for your meticulous efforts! 👍** 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._



Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadat3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
